CartoDB/cartodb
# Quota precheck

This PR implements https://github.com/CartoDB/camshaft/issues/238. This is a complete revamp of https://github.com/CartoDB/cartodb/pull/11297

We are using a dataservices api to: 
- obtain the user's quota
- check the quota before to run an analysis. 

Whenever we enter in (or changes) the analysis form view, we request the user quota in order to update the data. Take in mind that the quota could have shared under the organisation, so other users could change the remaining quota at the same time. So at least, one request is made. Also, when the analysis needs a quota precheck, two more requests need to be made:
- one for getting an estimation to get the credits needed to run the analysis
- another to check if the quota is enough to run the analysis with the estimation above.

So, if the analysis needs to check the quota, 3 requests are made.

### How to test
- Check the scenarios above, making sure that the request matches.
- You should check that the information we are showing to the user is right in these cases:
    - The user has soft limit but no quota assigned.
    - The user has soft limit but no quota remained.
    - The user has soft limit but the quota is not enough to run the whole analysis.
    - The user has soft limit and the quota is enough.
    - The same as above but with hard limit.
- Also, if dataservices api is not ready, you should check that analyses that require quota are not available. For instance, data observatory should not be part of the list of analysis in the modal.

Take this messages as a guidance:
![3f5888f4-e8a1-11e6-83a0-1f110c7b1ea8](https://cloud.githubusercontent.com/assets/1366843/22707804/43a98138-ed74-11e6-8076-6ef552d73340.png)

Test it in ded05, all has been setup properly there. Ask for user too.

cc @xavijam
